src/gazebo_vision_plugin.cpp

```diff
@@ -0,0 +1,173 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+ /**
+  * @brief Vision Plugin
+  *
+  * This plugin simulates VIO data
+  *
+  * @author Christoph Tobler <christoph@px4.io>
+  */
+
+#include <gazebo_vision_plugin.h>
+
+namespace gazebo {
+GZ_REGISTER_MODEL_PLUGIN(VisionPlugin)
+
+VisionPlugin::VisionPlugin() : ModelPlugin()
+{
+}
+
+VisionPlugin::~VisionPlugin()
+{
+  event::Events::DisconnectWorldUpdateBegin(_updateConnection);
+}
+
+void VisionPlugin::getSdfParams(sdf::ElementPtr sdf)
+{
+  _namespace.clear();
+  if (sdf->HasElement("robotNamespace")) {
+    _namespace = sdf->GetElement("robotNamespace")->Get<std::string>();
+  } else {
+    gzerr << "[gazebo_vision_plugin] Please specify a robotNamespace.\n";
+  }
+
+  if (sdf->HasElement("pub_rate")) {
+    _pub_rate = sdf->GetElement("pub_rate")->Get<int>();
+  } else {
+    _pub_rate = DEFAULT_PUB_RATE;
+    gzerr << "[gazebo_vision_plugin] Using default publication rate of " << DEFAULT_PUB_RATE << " Hz\n";
+  }
+
+  if (sdf->HasElement("corellation_time")) {
+    _corellation_time = sdf->GetElement("corellation_time")->Get<float>();
+  } else {
+    _corellation_time = DEFAULT_CORRELATION_TIME;
+    gzerr << "[gazebo_vision_plugin] Using default correlation time of " << DEFAULT_CORRELATION_TIME << " s\n";
+  }
+
+  if (sdf->HasElement("random_walk")) {
+    _random_walk = sdf->GetElement("random_walk")->Get<float>();
+  } else {
+    _random_walk = DEFAULT_RANDOM_WALK;
+    gzerr << "[gazebo_vision_plugin] Using default random walk of " << DEFAULT_RANDOM_WALK << " (m/s) / sqrt(hz)\n";
+  }
+
+  if (sdf->HasElement("noise_density")) {
+    _noise_density = sdf->GetElement("noise_density")->Get<float>();
+  } else {
+    _noise_density = DEFAULT_NOISE_DENSITY;
+    gzerr << "[gazebo_vision_plugin] Using default noise density of " << DEFAULT_NOISE_DENSITY << " (m) / sqrt(hz)\n";
+  }
+}
+
+void VisionPlugin::Load(physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  getSdfParams(sdf);
+
+  // Store model
+  _model = model;
+
+  _world = _model->GetWorld();
+  _last_time = _world->GetSimTime();
+  _last_pub_time = _world->GetSimTime();
+
+  // remember start pose -> VIO should always start with zero
+  _pose_model_start = _model->GetWorldPose();
+
+  _nh = transport::NodePtr(new transport::Node());
+  _nh->Init(_namespace);
+
+  // Listen to the update event. This event is broadcast every simulation iteration.
+  _updateConnection = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&VisionPlugin::OnUpdate, this, _1));
+
+  _pub_odom = _nh->Advertise<odom_msgs::msgs::odom>("~/" + _model->GetName() + "/vision_odom", 10);
+}
+
+float _corellation_time;
```

Comment:
@ChristophTobler this isn't doing nothing here right?

Reply:
no... fixed